pkg/ds/interfaces.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+// Package ds defines and implements a collection of data structures,
+// designed to hold cross-products of two Set structures in a succinct and canonical way.
+// The most important interface is Product. It is implemented by ProductLeft using injective
+// (one-to-one) mapping from sets to sets, where each key-value pair defines a complete cross
+// product of the two sets.
+package ds
+
+type Sized interface {
+	IsEmpty() bool
+	// Size returns the actual, full size of the set. For Product, it returns the number of pairs.
+	Size() int
+}
+
+type Comparable[Self any] interface {
+	Equal(Self) bool
+	Copy() Self
+}
+
+type Hashable[Self any] interface {
+	Comparable[Self]
+	Hash() int
+}
+
+type Set[Self any] interface {
+	Hashable[Self]
+	Sized
+	IsSubset(Self) bool
+	Union(Self) Self
+	Intersect(Self) Self
+	Subtract(Self) Self
+}
+
+// Product is a cartesian product of sets S1 x S2
+type Product[S1 Set[S1], S2 Set[S2]] interface {
+	Set[Product[S1, S2]]
+	Partitions() []Pair[S1, S2]
+	NumPartitions() int
+	Left() []S1
```

Comment:
shouldn't `Left()` return `Set[S1]` instead of `[]S1` ?
Shouldn't the projection on one dimension be just a set without order and with no duplicated elements?

Reply:
This sounds correct, however, there is no obvious way to return an empty set for the empty partition. It may need an argument to Left() and Right() just to handle this case.